File: rle.js
// rle.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');

// ANSI colors
const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const ESCAPE = '\\';

function compress(text) {
    if (!text) return '';
    let result = [];
    let i = 0, n = text.length;
    while (i < n) {
        let ch = text[i];
        let j = i + 1;
        while (j < n && text[j] === ch) j++;
        let count = j - i;
        if (count >= 3) {
            result.push(ch, ESCAPE, count.toString());
        } else {
            for (let k = 0; k < count; k++) {
                if (ch === ESCAPE) result.push(ESCAPE);
                result.push(ch);
            }
        }
        i = j;
    }
    return result.join('');
}

function decompress(text) {
    if (!text) return '';
    let result = [];
    let i = 0, n = text.length;
    while (i < n) {
        let ch = text[i];
        if (ch === ESCAPE) {
            if (i + 1 < n && text[i + 1] === ESCAPE) {
                result.push(ESCAPE);
                i += 2;
                continue;
            }
            if (i + 1 >= n) throw new Error('Unexpected end after escape');
            let char = text[i + 1];
            i += 2;
            let numStr = '';
            while (i < n && text[i] >= '0' && text[i] <= '9') {
                numStr += text[i];
                i++;
            }
            if (!numStr) throw new Error('Missing number after escape');
            let count = parseInt(numStr, 10);
            for (let k = 0; k < count; k++) result.push(char);
        } else {
            result.push(ch);
            i++;
        }
    }
    return result.join('');
}

function readInput(filename) {
    if (!filename || filename === '-') {
        // read from stdin
        return fs.readFileSync(0, 'utf-8');
    }
    return fs.readFileSync(filename, 'utf-8');
}

function writeOutput(filename, content) {
    if (!filename || filename === '-') {
        process.stdout.write(content);
    } else {
        fs.writeFileSync(filename, content, 'utf-8');
    }
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log(colorize('Usage: node rle.js compress|decompress [input] [output]', 'yellow'));
        process.exit(1);
    }
    const mode = args[0];
    if (mode !== 'compress' && mode !== 'decompress') {
        console.log(colorize('Invalid mode. Use compress or decompress.', 'red'));
        process.exit(1);
    }
    const inputFile = args.length >= 2 ? args[1] : undefined;
    const outputFile = args.length >= 3 ? args[2] : undefined;

    let data;
    try {
        data = readInput(inputFile);
    } catch (err) {
        console.log(colorize('Error reading input: ' + err.message, 'red'));
        process.exit(1);
    }

    let result;
    try {
        if (mode === 'compress') result = compress(data);
        else result = decompress(data);
    } catch (err) {
        console.log(colorize('Error: ' + err.message, 'red'));
        process.exit(1);
    }

    try {
        writeOutput(outputFile, result);
        if (outputFile && outputFile !== '-') {
            console.log(colorize('Result written to ' + outputFile, 'green'));
        }
    } catch (err) {
        console.log(colorize('Error writing output: ' + err.message, 'red'));
        process.exit(1);
    }
}

main();

File: rle.rb
#!/usr/bin/env ruby
# rle.rb
# encoding: UTF-8

require 'stringio'

# ANSI colors
COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

ESCAPE = '\\'

def compress(text)
  return '' if text.empty?
  result = StringIO.new
  i = 0
  n = text.length
  while i < n
    ch = text[i]
    j = i + 1
    j += 1 while j < n && text[j] == ch
    count = j - i
    if count >= 3
      result << ch << ESCAPE << count.to_s
    else
      count.times do
        result << ESCAPE if ch == ESCAPE
        result << ch
      end
    end
    i = j
  end
  result.string
end

def decompress(text)
  return '' if text.empty?
  result = StringIO.new
  i = 0
  n = text.length
  while i < n
    ch = text[i]
    if ch == ESCAPE
      if i + 1 < n && text[i + 1] == ESCAPE
        result << ESCAPE
        i += 2
        next
      end
      raise "Unexpected end after escape" if i + 1 >= n
      repeat_char = text[i + 1]
      i += 2
      num_str = ''
      while i < n && text[i] =~ /[0-9]/
        num_str << text[i]
        i += 1
      end
      raise "Missing number after escape" if num_str.empty?
      count = num_str.to_i
      result << repeat_char * count
    else
      result << ch
      i += 1
    end
  end
  result.string
end

def read_input(filename)
  if filename.nil? || filename == '-'
    $stdin.read
  else
    File.read(filename, encoding: 'UTF-8')
  end
end

def write_output(filename, content)
  if filename.nil? || filename == '-'
    print content
  else
    File.write(filename, content, encoding: 'UTF-8')
  end
end

if ARGV.empty?
  puts colorize("Usage: ruby rle.rb compress|decompress [input] [output]", :yellow)
  exit 1
end

mode = ARGV[0]
unless ['compress', 'decompress'].include?(mode)
  puts colorize("Invalid mode. Use compress or decompress.", :red)
  exit 1
end

input_file = ARGV[1]
output_file = ARGV[2]

begin
  data = read_input(input_file)
rescue => e
  puts colorize("Error reading input: #{e.message}", :red)
  exit 1
end

begin
  result = (mode == 'compress') ? compress(data) : decompress(data)
rescue => e
  puts colorize("Error: #{e.message}", :red)
  exit 1
end

begin
  write_output(output_file, result)
  if output_file && output_file != '-'
    puts colorize("Result written to #{output_file}", :green)
  end
rescue => e
  puts colorize("Error writing output: #{e.message}", :red)
  exit 1
end
